### ISSUE TITLE
Add sub-layouts option (Enhancement)

### ISSUE THREAD

**Issue by Blkx-Darkreaper** [nhmkdev/cardmaker#121]
Could you add a feature where you can have a base layout and then multiple derivative sub layouts that build upon the base, but only include those elements that are unique to the sub layout?

This way you could modify the base layout without needing to duplicate the changes

**Comment by nhmkdev** [comment]
I investigated this a bit in the past and decided on a slightly different approach: element references

You can copy 1+ elements from one layout and paste them into another layout as references (manual covers a bit more on them).

**Comment by MarkVabulas** [comment]
I have done some brainstorming on how this feature could be implemented without a massive overhaul of any of the underlying systems.  I have recently been trying to streamline the process of having sub-layouts because I have a main card with multiple sub-pieces which are all very similar but unique, and the process can be automated in a manner of sub-layouts.  I currently have it set up such that one layout with transparent backgrounds exports an image based on the spreadsheet, and then a second layout loads these images into the correct places.  (I use custom naming schemes to find them in the output folder.)

The weakness of the above approach is that if I want customizations based on the main layout to work in the sub-layout, I have to create many variations of identical parts, just with different colors, to use in the primary scheme.  This ends up with hundreds to thousands of extra images that I'm referencing only for the purposes of inserting them in another layout.  Additionally, there is a headache of having all of these extra entries in the spreadsheet, instead of just a series of primary ones and then individual ones where they should be unique.

My proposed solution to this, to make it work, is a shortcut for many parts of this which I imagine would be easy to implement behind the scenes.  Basically, one designs a layout as normal to serve as a sub-layout.  It can reference its own rows in the spreadsheet in order to have the necessary duplication of interests.  Then, on another layout, there is a new element type, we can call it "Layout Reference" for simplicity.  The "description" field would be the name of the other Layout for duplications. What this layout does: for each row of it's own spreadsheet, it passes all of its own data dictionary as additional entries to the sub-layout.  This means that if the sub-layout is referencing "@[SpecialColorA]", and SpecialColorA is a field in the primary layout, then the secondary/sub layout would receive that as well, and be able to use that field. For **EACH** iteration of the layout which contains the "Layout Reference" field, it would invoke the SubLayout as well, passing in the additional fields from the primary spreadsheet row.  Then after that SubLayout has iterated, the rest of the main layout runs.  The outputs from the SubLayout could be stored in the typical output folder, using the custom names defined in its own properties, as normal.  Each invocation of the primary layout, which would trigger the sublayout, could use the outputs from the sublayout as inputs to its own fields. 

The "Layout Reference" field wouldn't have a visual representation, just tell CardMaker which other Layouts to invoke for each iteration of the primary Layout, by passing the custom per-main-layout info along into the sub-layout's data definitions.

The use case that I have looks like this:
Main layout containing 11 sub-layouts which are definitions of different units available for the player.
Each of these sub-layouts is effectively the same, in terms of the content, except sometimes there are individual values which are unique to a specific unit. Additionally, each unit is supposed to be colored based on parameters unique to the invocation of the main layout.  Instead of creating 11 new lines in the sub-layout spreadsheet for each main layout, which is of course doable, it would be wonderful if the sub-layout only contained the special parameters for the unique variants, and then did lookups into the main layout for the coloration necessary to fulfill the main layout.  Sometimes the sub-layouts also have iconography which is associated with the main layout, but that would be the only differentiating element on the entire sublayout.  This icon could be defined in the primary layout, and then it would be automatically incorporated when invoked from the primary layout. 

In summary, TLDR, the requested functionality is merely to have a functionality where one layout can trigger the execution of another as a pre-step, and that the fields in the first layout are appended to each of the rows of the spreadsheet for the referenced layout.  We already have the wiring for the rest of the functionality as necessary, because of the benefits of transparent backgrounds, custom output names, etc.  

If you'd be open to it, I would be willing to try to implement this new field type, if you could tell me where you think these parts would go?  (New entry to the tab for `tabControl`, new layout on it to specify which layout to force to re-run in advance, and where to stick the code to append the data and force a pre-execution)

**Comment by nhmkdev** [comment]
I'll probably have to review this a couple times to make sure I'm understanding it all. We'll want a clear set of requirements to make sure we cover it all. 👍 

**Some things to consider**
* What index does the SubLayout process at with its own references? Is it always the same index as the PrimaryLayout? What happens if the counts do not match up (default to the last entry in the SubLayout)?
* I think making an Element that is purely a Layout Reference is **probably the easiest** (I hope 🤣) path. The only values that might impact how it is rendered would be the x/y position. This way a SubLayout could be re-positioned (maybe more could be nice).
* The main thing I would need to investigate is reference loading and translation so the PrimaryLayout reference data can be used as potential overrides when rendering/processing the SubLayout with its own reference. It's another layer of override that doesn't exist now.

**Clarification**
> The outputs from the SubLayout could be stored in the typical output folder, using the custom names defined in its own properties, as normal. Each invocation of the primary layout, which would trigger the sublayout, could use the outputs from the sublayout as inputs to its own fields.

I'm not quite sure I follow this. Is the idea that Rendering of the PrimaryLayout forces SubLayout(s) to export an image and the image is rendered?

**Repeating the concept in my own words** 
(just to make sure I'm understanding!)

* A Layout ("PrimaryLayout") can have SubLayouts
* SubLayouts should be drawn in order relation to other SubLayouts/Elements
* When rendering a given SubLayout it should have access to the Reference data of the PrimaryLayout. The data from the PrimaryLayout should **override** any data by the same name the SubLayout uses.
* One possibility would be to have the SubLayout render to an image and then draw that image when producing the PrimaryLayout.

I'll review again later to make sure I'm not missing anything.

**Comment by MarkVabulas** [comment]
Your repeat of things in your own words is complete and correct.  The piece that is missing will solve the requested clarification and also answer your first question about which indexes happen.  

- My suggestion is that the Referenced Layout is executed **in its entirety**.  So if the SubLayout has 15 entries, they are all executed.  This means that if someone wants a SubLayout with only 1 instance, that's fine.  It will be executed in each Layout that needs it. If someone wants a SubLayout that has many variations so they can include them all, then they get all of them processed, for every entry of the Primary Layout.  This isn't a duplication of work, it's just moving where it's done.  In my example that I'm trying to solve, I already had all of the same fields in the Primary Layout, duplicated 11 times, so I had hundreds of fields.  This merely cuts those down to a single Reference Layout field, which tells which SubLayout to execute.  Next answer tells how it's **USED**.
- The second and third answer is actually the same one: The SubLayout would create all of its assets **AS NORMAL**, the only change to how the SubLayout functions is that it has access to more/specific information coming from the Primary Layout.  
- This has the effect of creating an asset/output however the SubLayout is configured.  The part that comes up is that in order to USE the output, we create a Graphics entry like normal and put into it's description the filename(s) we expect to be in the output from the SubLayout.   Since we're guaranteed that the SubLayout is going to run before anything happens in the Primary Layout, all of the assets necessary would already be created. This also means that the assets can be created at a different resolution, and then stuck into the Primary Layout using all of the fancy features of the Graphics tag, including things like rotation, mirroring, scaling, shading, etc.
- I'm suggesting that the job of placing one or more assets from the SubLayout isn't the job of CardMaker, it's the job of the user to decide how the Outputs are used.  It's not as easily user friendly, but it's significantly more straightforward to design.  Forcing a user into a specific usage pattern takes away their choice and agency, and simultaneously requires more programming time on your part.   The goal of my suggestion is to make the processing merely happen to make the Layouts (be benefit of having access to the Primary Layout's data definitions), leave it up to the user if they want N new SubLayout output(s) and also how they want to use them.  You've done a great job of making the system flexible, and as complicated as it already is, the concept of using a SubLayout to make new assets for each Primary Layout, and then being able to use those generated assets through self-defined Graphics, is not overly complex.  Writing up an example is `"Primary Layout A needs a generated asset -> SubLayout B can generate that asset with parameters supplied by A. Then we just make a Layout Reference [A relies upon B], and a corresponding Graphics entry that looks for the outputs of B, and uses them accordingly."`

I am a strong supporter of having the Primary Layout definitions override any SubLayout definitions, if feasible.  I'm also perfectly happy forcing them to be unique and identifying them the way that conflicts are already identified, between th ings like the _defines files and normal columns.

Hopefully this has clarified what questions remained, and elucidated that the amount of work necessary to complete this task isn't actually immense.  We don't need to worry about coordinate systems, placement, forwarding of information, hard links, etc. We just need to ensure that the Referenced Layouts are executed _a priori_ and therefore have the generated assets available in time for the standard workflow.

**Comment by MarkVabulas** [comment]
It's important to mention that since the SubLayout is merely executing along it's normal output creation path, using its saved output parameters, the SubLayout's outputs can also be used without being in a SubLayout at all.  This would guarantee that the same asset being used in multiple places would be identical, especially if the SubLayout was being printed on something larger AND ALSO being printed as something individual like a companion card.

**Comment by MarkVabulas** [comment]
I would very much like the chance to pursue this as a test or partial implementation, if you could guide me in the direction of where you think the changes should be made.

**Comment by nhmkdev** [comment]
I think I'm slowly understating what you're going for with this. 🤔 

## Another attempt at my interpretation 😃 

### Layouts
* SubLayout (2 items)
* PrimaryLayout (10 items)
  * Has a reference that looks kind of like the table below

| count | title | color |
|--|--|--|
|1|Rock|Grey|
|1|Tree|Brown|

### Export of PrimaryLayout
1. Row 1 of the reference:
   1. Export the SubLayout with the row data fromPrimaryLayout Row 1 (`title` **Rock** and `color` **Grey** are passed in)
       1. Overwrites any existing **SubLayout_1.png** and **SubLayout_2.png** (uniquely generated with the `title` & `color`)
   2. Rest of the elements on PrimaryLayout are rendered and exported to **PrimaryLayout_1.png** (some might use the **SubLayout_?** files)
2. Row 2 of the reference:
   1. Export the SubLayoutX with the row data fromPrimaryLayout Row 2 (`title` **Tree** and `color` **Brown** are passed in)
       1. Overwrites any existing **SubLayout_1.png** and **SubLayout_2.png** (uniquely generated with the `title` & `color`)
   2. Rest of the elements on PrimaryLayout are rendered and exported to  **PrimaryLayout_2.png** (some might use the **SubLayout_?** files)

... continues for all 10 items in PrimaryLayout.

## Additional Thoughts

* One concern is that the references will reload every time you re-export the SubLayout (if you happen to be using Cached Google References you might be avoiding that).
* I've hit some occasions where the operating system reports files as missing so we'd need to be sure the file is completely written to disk and available before the PrimaryLayout attempts to use it via a Graphic element.

## Alternative (sort of) approach
I think offering parameterized layouts via a Graphic element could be an interesting approach.

`::SubLayout::[Card Index]::[parameter name]::[parameter value]::[parameter name]::[parameter value]::(continue with more parameters)`
 
This would force the SubLayout to load, pass along the specified parameters, generate exactly the card index specified and potentially do so completely in-memory.

## Files To Investigate (specifically around exporting layouts)

Files to potentially look into for the export of the SubLayout:
* `CardMaker\Card\Export\FileCardExporter.cs`
* `CardMaker\Card\Export\CardExportBase.cs`
* `CardMaker\Card\Deck.cs`

I think you will need to pass in an override set of reference data to the `FileCardExporter` and then merge/override the values on the `Deck` (ValidLines and the Defines). 

It's also possible you could perform the merge/override in the `CardMaker\Card\DeckReader.cs` itself.

The biggest overall challenge is getting whatever parameters you need into the SubLayout export.

**Comment by MarkVabulas** [comment]
I really love the Alternative approach suggested, because that would solve having to make an entirely new element type.  My concern is that the card index might be strange, because it will change over time, and I would end up with `#(if @[Card Override 15] != 3 then @[Card Override 15 else 3 )#` in every field, and then if just one card gets inserted into the SubLayout's spreadsheet then every override and every default would also have to change for every field and every override field in the PrimaryLayout spreadsheet.

My plan for just adding the override fields from the PrimaryLayout and the following the rules for the SubLayout itself means that it can output Custom_@[Unique-But-Constant Card ID].png for each of it's cards, and then the PrimaryLayout just needs to have a Graphics field which specifies Custom_@[Unique-Card-To-Load].png, and part of the logic for the Graphics field would have an if-statement which would default to one of the base layouts if nothing is specified for the Primary Layout. Making it depend on specific indices in the SubLayout complicates these matters from an automation standpoint, we can no longer just "create cards", we have to track the card dependencies in our heads too.  Allowing for a range of cards to be accessed through the suggested alternative approach, so a range could be created, would somewhat fix this, but then we risk causing a re-interpretation for the entire sub-layout for every Graphics element which wants N potential imports from the SubLayout.  10 Graphics elements, each one needing the entire SubLayout, even though the outputs of the SubLayouts are all the same because they're fed from the same PrimaryLayout row, means that we would need better dependency tracking.

As far as ensuring that the files are written in SubLayout before grabbing them from PrimaryLayout:  I believe that closing the file handle causes a flush and in this case it would function as we want.  Adding a file-availability check to PrimaryLayout as a bug fix shouldn't be too difficult if this error crops up.

As for the concern of re-loading the references repeatedly, I wasn't aware that you're force reloading the references at the start of each layout running.  My suggestion would be to step which does a refresh at the start of the process, then keeps the entire layout's data of everything in memory to reuse for the rest of the layout.  Basically, allow a SubLayout to NOT re-check its data after the initial grab.  In my case I'm testing with Excel, so the testing phase can be low-impact for now.

In summary, yes your idea about how the flow works is correct now.  You do fully understand.  The key you were missing which unlocks simplified access is that we can have custom filenames for SubLayout, by using the "Name Format" field for the SubLayout.  So instead of SubLayout_N.png, we get Special_Card_Orange.png or Default_Card_Ally.png (from the same SubLayout invocation!).  Perhaps in the Layout Reference field we can have `[SubLayoutName]::[SubLayout File Name Format]` if we want to use that functionality to give us custom filenames on a per-layout basis.  That would be pretty effective.  This SubLayout File Name Format would be easy to understand as well, since the fields in it can use the full Incept translator and will have access to the SubLayout fields as well as the Primary Layout's fields.

**Comment by MarkVabulas** [comment]
Correction, I properly have one layout causing the execution of another layout, based entirely upon a new ElementType which I created called "SubLayout".  The element 'value' is the name of the new Layout.  

- To simplifying development, since I'm not sure how you want to handle it, I added a flag to disable ProgressReporting since it's no longer consistent.  
- I added a function to ProjectManager called GetLayoutIndex which takes the name of a Layout and returns it's index. 
- New ElementType::SubLayout
- Function inside Deck that combines a new field ParentPrintLine with the CurrentPrintLine.  The ParentPrintLine is set as the value of this same function, to allow for recursive nesting of layouts.  All instances of CurrentPrintLine in the code use this new function. **_THIS IS WHERE ITS INCOMPLETE_**  I need help figuring out how to make this function properly combine the lines, and their corresponding column names (I guess they're only stored in Translator?)
- All other changes are in FileCardExporter.  I created a new function which traverses the layout in advance and isolates each of the SubLayouts, in forward order.  It uses the Translator of the CurrentPrintLine+ParentPrintLine on the value, then does a search for the result inside ProjectManager.Instance.GetLayoutIndex() referenced earlier. Any layouts it finds get added to a list and returned.
- Inside FileCardExporter::ExportThread, the only changes I needed to make were to figure out the SubLayout list before we loop through the cards, and then at the start of the card loop, create a fresh FileCardExporter of similar/default settings for each of the found SubLayouts, and then execute it in-thread.  After that loop is finished, it resets the current Layout through ChangeExportLayoutIndex(nIdx) to bring consistency back to the Parent Layout and continues with that particular card.  

All of the code referenced above is relatively trivial to write.  I've pushed all of my changes to my fork at [https://github.com/MarkVabulas/cardmaker](https://github.com/MarkVabulas/cardmaker/commit/3b83c967605f9d0d1d808b5c2726825c2c84cf96) if you'd like to see the diff for the entire change log.  I'm curious what you think about this stuff.  I'll create a PR if you would like, but I'm pretty sure none of this is up to the standards or methods you might want to use in the end.

**Comment by nhmkdev** [comment]
I'm glad you were able fairly far along.

I'll add some comments to the submit you made on things I found when trying to get things somewhat working.

I'll take a more in-depth look hopefully in the next day or so.

**Comment by nhmkdev** [comment]
Added a comment and patch with my in-progress hacking: https://github.com/MarkVabulas/cardmaker/commit/3b83c967605f9d0d1d808b5c2726825c2c84cf96#commitcomment-140690353

**Comment by MarkVabulas** [comment]
New seemingly-100% working commit.  Testing required, but I'm going to do that extensively, hopefully this weekend.  The commit changelog explains what I did.
[commit 4bf9655](https://github.com/MarkVabulas/cardmaker/commit/4bf96556b5023ba6ba5df027460756773d0ca443)

I incorporated your changes and made the necessary improvements.  It all works correctly now, including with the cache.  Each Primary Layout is triggering a rebuild of the different cards in the SubLayout and able to access the fields in the SubLayout.  I removed the GetFullPrintLine since it actually became useless once the Translator was incorporated.  The Primary Layout doesn't/shouldn't need access to the SubLayout, and the SubLayout only needs the fields for translation purposes.  I also saw your `#warning` at Deck.cs:264, but I don't think we actually need that info anywhere.  What were you referring to?

I did switch the order of checking for the Translator to use the PrimaryLayout info first, so it can function as an override.  I also reverted the not-found functionality back to as-before, since it shouldn't silently fail for a Layout just because it's also going to be a PrimaryLayout.  We should have the fields defined in both, if it's being used, AND overridden.

I'm really excited to try this on my problem set, but today is a day to relax :)

**Comment by MarkVabulas** [comment]
And sorry for the line-endings getting screwy, I tried to make sure they wouldn't get messed up this time and it still happened.

**Comment by MarkVabulas** [comment]
![image](https://github.com/nhmkdev/cardmaker/assets/1007185/238d2130-492e-4f30-be3a-db21f10f8e3d)
So this works great.  The sub-layouts are all originally defined as white, but the purple ships and red boxes around the numbers are all re-defined in the ParentLayout as overrides.  The issue we are having is that it's reusing the image cache from the first card on subsequent ones, even though it is still correctly generating the SubLayouts (I can see them in File Manager thumbnails getting updated, but then the output for the next card contains the cached images).

**Comment by MarkVabulas** [comment]
It should be stated that (I didn't time it) this card appears the take an essentially equal amount of time.  I kept my old layout, and it takes about 1 minute.  This new one using SubLayouts takes about a minute as well.

I included the ImageCache.ClearImageCaches() call inside the loop after resetting the Primary Layout CardPrintIndex, and it works perfectly now.  [link to commit](https://github.com/MarkVabulas/cardmaker/commit/19b62fd83177fd2cb2f9e4240b4881bd99d74d31)

**Comment by MarkVabulas** [comment]
![image](https://github.com/nhmkdev/cardmaker/assets/1007185/70d91161-ee79-45bd-b254-ddc3ba7bdd64)
![image](https://github.com/nhmkdev/cardmaker/assets/1007185/1732fe9b-99bf-4051-b170-c23abdd79282)

The only difference between all of this mini-ship cards is what is defined in the PrimaryLayout.

**Comment by nhmkdev** [comment]
It definitely seems like there is a bug that is requiring the call to `ChangeExportLayoutIndex(nIdx);` on every card export. I'll have to check out why that is necessary. The primarylayout/sublayout exporters and their CurrentDecks should be independent (but may not be!).

## New Element Type vs. Layout Setting
For this initial implementation (not the advanced one that would use a Graphic element type): Do you need need the sublayout name to be translated as a completely independent element? I am considering just making a layout level variable like `sublayoutsExportPerCard` and it would just have a comma delimited list (`shipbackground,gamelogo,base`).

**Comment by MarkVabulas** [comment]
`ChangeExportLayoutIndex `is required because the call to CardRenderer and all all calls inside there use the (GLOBAL!!!!!) CurrentDeck and the deck changes in the start of every call to ExportThread.  (CurrentDeck is fully global, from what I can tell.)  Since we're calling ExportThread from inside ExportThread, the Sublayouts are overwriting the deck which was chosen by the PrimaryLayout.  You don't really have a full MVC style separation of data happening because your renderer is using whatever some current state is, which is shared elsewhere, instead of using only parameters passed in from elsewhere.  If you wanted this to function better, then CardRenderer, and even deeper the exporter, should only access data from things passed into them.  That has knock-on effects in places like the Translator because it's also using the global state.  Basically, everywhere in the software you've saved yourself the hassle of passing around the deck/layout because you only ever needed 1 and only 1 at any given time.  If we wanted this to be more seamless, it would all have to be thrown around as arguments instead then we should ship off commands like SubLayout renders in a multi-threaded way, because they wouldn't all simultaneously be using the CurrentDeck as the central repository of what's happening.  Just the mere existence of using a global CurrentPrintLine is an example of a design which would make this part of the project difficult.

For the part about New Element Type vs Layout Setting, I had the same thought, and the only conclusion I came to was that there are actually 3 ways to do it. 

1. As we currently are, with a SubLayout Element (Which has the benefits of a full implementation of the incept translator for the value field and selectively choosing layouts programmatically)
2. Using a Layout-wide property which we can set as "we need ____"
3. As a PER-ELEMENT reference which defines which other layout this Element itself specifically depends upon

- The first choice I feel is very flexible because of the incept translator, the ability to double-click to disable the SubLayout which would automatically disable the execution of it (if we don't need it every time), and it's also visually evident that it's there as a process/step because it's in the list of elements.  I'm fully expecting and hoping to rely on the existence of the incept translator for SubLayout selection.
- Having a layout-wide property is great because it more closely represents the relationship between two different layouts, and it is technically a properly of the layout moreso than as an element itself.
- Having a per-element reference, for Graphics AND textual fields, that declares that the contents depend directly upon a specific Layout for that element to render correct gives the flexibilty of disabling Layouts, but also means we can know more directly in the future which parts rely on what.  So if a specific element's value is changed, we could theoretically update the SubLayout to make sure it renders correctly, without having to do that same processing task Layout-Wide during any given edit.  The reason that even FormattedText fields might need a SubLayout to execute first is because of inline images or background images, etc.

The one thing which is currently missing from the SubLayout implementation is a more robust preview functionality, but in my view, the base SubLayout would need to exist as a pre-rendered item first anyway just to mock it up.  I feel it's not unreasonable, and would be understandable to the user, that before a SubLayout can be used in a Primary Layout that it needs to be rendered first at least once, independently.  While the SubLayout's exported preview-contents might not end up matching the final output, a quick run of a single-card export followed by a cache refresh gives immediate results for previewing the final task as-is, without forcing the user to sit through a full SubLayout re-export just to design their Primary Layout.

In the end, if we can keep the text translator working on the contents of whatever field is used for the SubLayout, I think that's optimal.  Personally, I'm not wed to any given solution, but I really like the idea of having an Element which serves as a visual reminder and controller for SubLayout References, it makes it feel more obvious that "yes, this is going to have to do something" and I'm worried that hiding that behind a settings screen might make it feel less impactful than it really is.  SubLayouts ARE "big-ask" types of card rendering.

**Comment by nhmkdev** [comment]
> `ChangeExportLayoutIndex `is required because the call to CardRenderer and all all calls inside there use the (GLOBAL!!!!!) CurrentDeck and the deck changes in the start of every call to ExportThread. (CurrentDeck is fully global, from what I can tell.) 

CurrentDeck is not actually global (fortunately!). 😄 But, it's possible there are some other global-like aspects that may cause bugs with sublayout exports. The loaded project is definitely global, but the card rendering should be independent. You could have 20 layers of exporters and they shouldn't interfere with each other (at least that was the plan long ago).

To further clarify how CurrentDeck is constructed for exporters -- Each FileCardExporter extends CardExportBase. CardExportBase has this as part of construction:
```
            CardRenderer = new CardRenderer
            {
                CurrentDeck = new Deck()
            };
```
The intention was to allow the application UI to render and maintain its own CardRenderer/CurrentDeck (in CardCanvas and every exporter). 

## Something to try:

Can you let me know what happens with this change in FileCardExporter with your export of sublayouts? This is working for me so I want to narrow down where things aren't working for your use case. My test case is fairly trivial but I am able to export the primary with sublayout exported images that are customized by primary reference values.

```
if (vSubLayoutArray.Count() > 0)
{

//this call should not be necessary and is the cause of needing the CurrentDeck.CardPrintIndex assignment below
	//ChangeExportLayoutIndex(nIdx);

// this should be harmless but also not necessary 
	//CurrentDeck.ResetDeckCache();

// as mentioned above this is caused by the ChangeExportLayoutIndex
	//CurrentDeck.CardPrintIndex = nCardId;

// definitely necessary as the exported file names of previously loaded files could be the same
	ImageCache.ClearImageCaches();
}
```

A potential TODO item would be for me to adjust the image cache to store the last updated time so we can avoid a complete cache wipe with `ImageCache.ClearImageCaches();`

**Comment by MarkVabulas** [comment]
So I removed those lines and it seems to function as before, with no discernable difference.  

One issue that cropped up is that occasionally, randomly, and without logic, about 5% of the time ALL of the defintions inside the SubLayout fail.  The images still load, and they are controlled by the Incept translator so the values aren't completely nulled out, but in general every single SubLayout fails for that particular Primary Layout.  Then if I re-run that individual card solo using the Export button on the card, it works fine.  But it's not guaranteed that it fails and it's not consistent which one it happens to during the run.

**Comment by MarkVabulas** [comment]
I'm having trouble reproducing it to show you an example, but it has happened multiple times, and each time I confirmed it could be fixed, but I'm having trouble getting it to fail on command (it knows I'm watching!).

**Comment by MarkVabulas** [comment]
![image](https://github.com/nhmkdev/cardmaker/assets/1007185/a0c7674d-7a03-409b-b57a-72cc51a60839)
Here we can see what the SubLayout ends up producing.  Clearly the FactionIcon is coming from the ParentLayout, and showing up on the SubLayouts, as well as the background correctly, but the rest of the SubLayouts are broken.

**Comment by MarkVabulas** [comment]
My newest commit to my main branch includes FileCardSingleExporter working as well.

**Comment by nhmkdev** [comment]
I've seen some odd cases over the last few months where a CSV file reference wouldn't be found on disk (the sample.csv which has existed on my PC for years). Definitely odd.

Are you seeing any log messages about files failing to load? With the ProgressReporter removed some issues with the sublayout export might be logging to nowhere. Sublayout exports might need a simple IProgressReporter implementation to at least allow for `AddIssue` to be called and tracked.

I might have to programmatically call the sublayout export a 100 or 1000 times in row and watch for issues.

Let me know once you feel like you're in a stable spot and I'll start planning the work on bringing things over to the main app repo. I'm probably going to stick with the element type approach and potentially do other things with it in the future.

**Comment by MarkVabulas** [comment]
I didn't notice any particular log messages or errors.  We'll have to investigate, but I can run it 10 times in a row with SubLayouts and not have it fail even once, so it's truly intermittent.

I am currently going through and doing more extensive testing of the SubLayouts, but I'm currently not foreseeing any more changes to the system than what we've already got in my HEAD commit on my fork. (The latest commits, I force-untabified my files before committing them back in)

**Comment by MarkVabulas** [comment]
Would you like me to submit a PR with my changes?

**Comment by MarkVabulas** [comment]
Concerning the IProgressReporter, if I had more familiarity with it, I would have implemented it as a 4th Progress Bar on the Export window, to show the progress when going through any sublayout(s).  That would allow for visualization, but we'd also have to keep track of the levels of ProgressBars deep that we are, in case there are nested SubLayouts.

**Comment by nhmkdev** [comment]
Initial submission here: https://github.com/nhmkdev/cardmaker/commit/92c6eb507f8f561f1a8c6eea7be9980cad1e29df

I'm going to close out this ticket and migrate to https://github.com/nhmkdev/cardmaker/issues/141
